feat: simple in-memory thread-safe cache

ChBLA opened the pull request:
Adds a simple in-memory thread-safe cache, with scheduled garbage collection.

KimNorgaard commented:
It's a good idea. I'll give it a review but it will probably be a few days before I get to it.

KimNorgaard commented:
So in general I'd say it's a good thing to have a solid in-mem type safe cache here. Or rather, I'd like there to be more (future) options which the current implementation does not easily support.

I think the implementation needs to be split up as it currently covers a many concerns together (ttl, gc, cache items). For a `cache` package I would expect the `Cache` interface to be very small and very general (Set, Get, Delete, maybe Clean and Len). It's fine to have specific cache implementations that implements that interface - one of which could be an in memory one (or safeMap if we want to keep transparency). I'd advise that we keep the implementations as specific as possible (in-mem could or could not use a map for instance, so that would be two different implementations - it could or could not support ttl, so that would be another way to split it up). All of the features are implementation details to the various implementations. They should not pollute the general interface, IMHO.

I also don't think we necessarily need all of the functions currently implemented here. Do we have use cases that need fine grained control over ttl per cache item and/or garbage collection intervals? If so I'd make specific implementations and/or interfaces for those as it seems like a lot to put into a single "simple" caching implementation.

KimNorgaard commented on cache/cache.go:
```diff
@@ -0,0 +1,443 @@
+	items map[string]Item
+	// mu provides concurrency control for safe access to items.
+	mu sync.RWMutex
+	// garbageCollector manages the periodic removal of expired items.
+	garbageCollector *garbageCollector
+	// cacheInterval specifies the default duration for items added without a specific TTL.
+	cacheInterval time.Duration
+	// garbageCollectorInterval defines how often the garbage collector runs.
+	garbageCollectorInterval time.Duration
+	// allowOverwrites determines if Add operations can overwrite existing, non-expired keys.
+	allowOverwrites bool
+	// clock provides an abstraction for time, useful for testing.
+	clock Clock
```
I would give serious consideration to these features. Do we really need to have that level of control? `cacheInterval`/`ttl` is probably fine, but I just don't know about the others. It doesn't go well with the 'simple' label.

KimNorgaard reviewed (commented):
The code quality is good but I'm not completely sure about the scope and organization of the package. I've left some comments to elaborate.

ChBLA commented:
I tried doing as suggested and minimized the scope of the interface, moved the (previous `cache` now `safeMapCache`) to its own package, and reduced the flexibility of the previous options. 

I was not exactly sure on how best to allow a consumer to define a TTL for the elements. Having the TTL in the input object felt forced, as all values either received or returned must then be accessed to get the actual value. Using an option did not seem appropriate either as the interface definition would have to change and I could not find a unified "SetOption" usable for all future cache implementations. I ended up simply adding a `SetWithTTL` function to the `safeMapCache` (not the interface) to allow for a custom TTL for a given item. 

I removed all unnecessary methods (i.e. Add) and reduced the footprint needed for garbage collection. Currently, the cache must implement one method `deleteExpired` to allow for garbage collection. Both `safeMapCache` and `garbageCollector` use their own `Clock` interface to allow for testing. This was the minimal amount of "testing" code that I could find that would still allow for proper testing independent of the system time.

KimNorgaard reviewed (commented):
I think this is good for a simple implementation but there are some things that could be reconsidered.

KimNorgaard commented:
We talked about using capability interfaces like this one:

```go
type TTLCache[K comparable, V any] interface {
    Cache[K, V]
    SetWithTTL(ctx context.Context, key K, value V, ttl time.Duration) error
}
```

KimNorgaard commented on cache/inmem/safemap.go:
```diff
@@ -0,0 +1,260 @@
+// It initializes the items map and sets up the garbage collector.
+// The garbage collector is started in a separate goroutine.
+// The clock is set to the real time package by default.
+func NewSafeMap[K comparable, V any]() *safeMapCache[K, V] {
```
Should this implement at default ttl that Set uses?

ChBLA commented on cache/inmem/safemap.go:
```diff
@@ -0,0 +1,260 @@
+
+// safeMapCache is a thread-safe in-memory cache implementation.
+// It uses a map to store items and a mutex to ensure thread safety.
+type safeMapCache[K comparable, V any] struct {
```
For now, I'll simple set a static limit and then reject new items once that limit is reached. Other policies can later be implemented either in this cache implementation or others.
